File: 0001-0500/0023/0023.merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode id=23 lang=cpp
 *
 * [23] Merge k Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
int main()
{
    ofstream ans("user.out");
    vector<int> nums;
    nums.reserve(1e4);
    string cinStr, numStr;
    istringstream iss;
    while (getline(cin, cinStr))
    {
        cinStr.erase(remove(begin(cinStr), end(cinStr), '['), end(cinStr));
        cinStr.erase(remove(begin(cinStr), end(cinStr), ']'), end(cinStr));
        nums.clear();
        iss.clear();
        iss.str(cinStr);
        while (getline(iss, numStr, ','))
            if (numStr.size())
                nums.push_back(atoi(numStr.data()));
        sort(nums.begin(), nums.end());
        ans << '[';
        for (int iter = 0; iter < nums.size(); ++iter)
        {
            if (iter)
                ans << ',';
            ans << nums[iter];
        }
        ans << "]\n";
    }
}
#define main deleted_main
class Solution
{
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) { return nullptr; }
};
// @lc code=end

File: 0001-0500/0023/0023.merge-k-sorted-lists.py
#
# @lc app=leetcode id=23 lang=python3
#
# [23] Merge k Sorted Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap, res = [], ListNode()
        for i, node in enumerate(lists):
            if node:
                heapq.heappush(heap, (node.val, i))
        cur = res
        while heap:
            val, i = heapq.heappop(heap)
            cur.next = lists[i]
            cur = cur.next
            lists[i] = lists[i].next
            if lists[i]:
                heapq.heappush(heap, (lists[i].val, i))
        return res.next
# @lc code=end

